### PR TITLE
test(firezone-tunnel): create a tunnel device in unit tests during CI

### PR DESCRIPTION
**Pull request by ReactorScram**
This catches two of the mutants, according to `cargo-mutants`.

~~Unfortunately since `cargo test` runs in one process, it's all-or-nothing for sudo, this will run all unit tests as sudo.~~
(This explanation is not exactly correct, `cargo test` does run _a_ subprocess, but still, there is no way to request sudo or non-sudo runners for specific tests, since it's just an environment variable, and since many tests run in parallel in different threads of the same process.)

Here it is passing in Linux: https://github.com/firezone/firezone/actions/runs/8382799272/job/22957555987#step:5:3160

And Windows: https://github.com/firezone/firezone/actions/runs/8382799272/job/22957558003#step:5:1006

```[tasklist]
### Before merging
- [x] Try `#[ignore]` attribute
- [x] Fail gracefully if `sudo` isn't available
```

### CODE REVIEW

**Comment by conectado** [comment]
I don't think we should require `sudo` in tests. In general, if anything like that is require I think we should use `#[ignore]` for that test but as a general rule I don't like using `cargo test` to test IO.

**Comment by ReactorScram** [comment]
@conectado Because it would leave changes on the filesystem / other global mutable state of the dev system running the tests?

I could put it in a separate test script or something, but then I can't use `cargo mutants` to find what code is covered and not covered.

**Comment by conectado** [comment]
> @conectado Because it would leave changes on the filesystem / other global mutable state of the dev system running the tests?
> 
> I could put it in a separate test script or something, but then I can't use `cargo mutants` to find what code is covered and not covered.

I prefer skipping cargo-mutants, IO can be something flaky and invasive to test indeed and `cargo test` should be quick and unobtrusive. We can have an example and run it as part of CI.

**Comment by conectado** [comment]
@ReactorScram can `cargo-mutants` be hooked into commands other than test?

**Comment by ReactorScram** [comment]
@conectado There might be a trick we can do. Looking at https://mutants.rs/nextest.html and https://mutants.rs/cargo-args.html

Maybe we could stick it behind a crate feature or something, and have that only enabled for mutants using `mutants.toml`? Or even just an env var, and leave a comment saying, "This test is skipped unless you set FIREZONE_DO_COMPLICATED_IO_TESTS, which cargo mutants  and some CI jobs will always set" and have it instantly pass if that env var isn't set.

I guess a containment crate wouldn't work, since `cargo test --all` or `--workspace` would pick it up.

**Comment by conectado** [comment]
> @conectado There might be a trick we can do. Looking at https://mutants.rs/nextest.html and https://mutants.rs/cargo-args.html
> 
> Maybe we could stick it behind a crate feature or something, and have that only enabled for mutants using `mutants.toml`? Or even just an env var, and leave a comment saying, "This test is skipped unless you set FIREZONE_DO_COMPLICATED_IO_TESTS, which cargo mutants and some CI jobs will always set" and have it instantly pass if that env var isn't set.
> 
> I guess a containment crate wouldn't work, since `cargo test --all` or `--workspace` would pick it up.

If we want to do this I think the [`#[ignore]`](https://doc.rust-lang.org/reference/attributes/testing.html#the-ignore-attribute) attribute is our best bet since it's designed for that and then use `--include-ignored`

File: .github/workflows/_rust.yml [hunk 1]
```diff
@@ -55,16 +55,19 @@ jobs:
           - runs-on: macos-14
             packages: -p connlib-client-apple -p snownet
           - runs-on: windows-2019
-            packages: -p firezone-gui-client -p connlib-client-shared -p snownet
+            packages: -p connlib-client-shared -p firezone-gui-client -p firezone-tunnel -p snownet
           - runs-on: windows-2022
-            packages: -p firezone-gui-client -p connlib-client-shared -p snownet
+            packages: -p connlib-client-shared -p firezone-gui-client -p firezone-tunnel -p snownet
```
Comment (ReactorScram) [inline .github/workflows/_rust.yml:60]: Alphabetized them while adding `firezone-tunnel`

**Comment by ReactorScram** [comment]
@conectado `#[ignore]` worked fine

**Review by jamilbk — COMMENTED** [review]
LGTM but will defer to @conectado

**Review by conectado — APPROVED** [review]
In the long-term I'd like to have the `Tunnel` be IO agnostic, so we can create it without depending on the `tun` device.

**Comment by ReactorScram** [comment]
Agreed, though I'd also like to test the I/O in VMs automatically since I/O is where all the painful bugs are. If we don't test it somehow, it gets 0% coverage by default.

**Pull request merged**